### PR TITLE
[Clang] Add explicit visibility symbol macros

### PR DESCRIPTION
**Pull request by fsfod**
This is part of the effort to support for enabling plugins on windows by adding better support for building llvm and clang as a DLL.  These macros are similar to the ones i added in #96630, but are for clang.
Added explicit symbol visibility macros definitions that are defined in a new header and will be used to for shared library builds of clang to export
symbols. 
Updated clang cmake to define a macro to enable the symbol visibility macros and explicitly disable them for clang tools that always use static linking.

### CODE REVIEW

File: clang/tools/libclang/CMakeLists.txt
Comment (compnerd) [inline clang/tools/libclang/CMakeLists.txt:169]: This will override the `_CINDEX_LIB_` symbol, so this seems incorrect.

File: clang/cmake/modules/AddClang.cmake
Comment (compnerd) [inline clang/cmake/modules/AddClang.cmake:112]: Why is this being added?

Comment (fsfod) [reply clang/cmake/modules/AddClang.cmake:112]: This is the way clang component libraries are indirectly defined, this commit sort of related https://github.com/fsfod/llvm-project/commit/f539bdcef3d299e7ea658c4575b860a082bcda71 and this flang commit sort of explains it as well https://github.com/llvm/llvm-project/commit/424ad163c715e478d37cfd1efaae148cdd6aa815

File: clang/tools/libclang/CMakeLists.txt
Comment (fsfod) [reply clang/tools/libclang/CMakeLists.txt]: I've fixed this and changed it to CLANG_BUILD_STATIC as well as adding a comment why its needed

File: clang/include/clang/Support/Compiler.h
Comment (AaronBallman) [inline clang/include/clang/Support/Compiler.h:69]: Please add a newline to the end of the file.

**Review by AaronBallman — COMMENTED** [review]
One thing I don't understand about this PR is why we need Compiler.h in Clang -- wouldn't the LLVM definitions in their Compiler.h work for Clang as well? (This would probably be worth explaining in the patch summary.)

**Comment by fsfod** [comment]
> One thing I don't understand about this PR is why we need Compiler.h in Clang -- wouldn't the LLVM definitions in their Compiler.h work for Clang as well? (This would probably be worth explaining in the patch summary.)

We could if we didn't need the visibility macros in different states of dllexport and dllimport in the same translation unit for like LLVM symbol to be import and a Clangsymbol to be exported.

**Comment by compnerd** [comment]
> One thing I don't understand about this PR is why we need Compiler.h in Clang -- wouldn't the LLVM definitions in their Compiler.h work for Clang as well? (This would probably be worth explaining in the patch summary.)

The symbol lookup in PE/COFF is two level and symmetric, bound to the module. As a result, each module needs to explicitly specify the ABI interface (contract). However, given that the interface is annotated differently (`__declspec(dllexport)` for the producer, `__declspec(dllimport)` for the consumer), we need to flip the definition. When doing so, we need to know which module is being built. We cannot just use the conjunction of the module identifier to select between the two variants as there may be interdependencies with dynamic linking (e.g. `#if defined(A_ABI) || defined(B_ABI)` doesn't work if B depends on A and dynamically links to A).

I do agree that we could put this in the commit message, though, this is likely something that deserves a note in the developer documentation because I fear that other developers are going to have this question repeatedly once this work is completed. The biggest point of contention with this work was the fact that it involves an ongoing maintenance cost as the ABI boundary is now being more rigidly defined which is/was not needed to support it on Unix due to the (global) single level binding.

**Comment by vgvassilev** [comment]
> > One thing I don't understand about this PR is why we need Compiler.h in Clang -- wouldn't the LLVM definitions in their Compiler.h work for Clang as well? (This would probably be worth explaining in the patch summary.)
> 
> The symbol lookup in PE/COFF is two level and symmetric, bound to the module. As a result, each module needs to explicitly specify the ABI interface (contract). However, given that the interface is annotated differently (`__declspec(dllexport)` for the producer, `__declspec(dllimport)` for the consumer), we need to flip the definition. When doing so, we need to know which module is being built. We cannot just use the conjunction of the module identifier to select between the two variants as there may be interdependencies with dynamic linking (e.g. `#if defined(A_ABI) || defined(B_ABI)` doesn't work if B depends on A and dynamically links to A).
> 
> I do agree that we could put this in the commit message, though, this is likely something that deserves a note in the developer documentation because I fear that other developers are going to have this question repeatedly once this work is completed. The biggest point of contention with this work was the fact that it involves an ongoing maintenance cost as the ABI boundary is now being more rigidly defined which is/was not needed to support it on Unix due to the (global) single level binding.

@fsfod, can you capture this comment in the commit message and the documentation as suggested?

File: clang/include/clang/Support/Compiler.h
Comment (compnerd) [inline clang/include/clang/Support/Compiler.h:34]: I don't know about this - we could be building LLVM dynamically and clang statically. I don't think that we should use the same macros for clang.

File: clang/include/clang/Support/Compiler.h
Comment (compnerd) [inline clang/include/clang/Support/Compiler.h:67]: The lack of indentation makes this very difficult to read. There are too many branches between the static and dynamic builds, and the various differences in the annotations.

File: clang/tools/amdgpu-arch/CMakeLists.txt
Comment (compnerd) [inline clang/tools/amdgpu-arch/CMakeLists.txt:11]: Why is this not allowed? Can we add a note about this?

File: clang/tools/clang-linker-wrapper/CMakeLists.txt
Comment (compnerd) [inline clang/tools/clang-linker-wrapper/CMakeLists.txt:29]: Similar

File: clang/tools/nvptx-arch/CMakeLists.txt
Comment (compnerd) [inline clang/tools/nvptx-arch/CMakeLists.txt:10]: Likewise

File: clang/include/clang/Support/Compiler.h
Comment (fsfod) [reply clang/include/clang/Support/Compiler.h:67]: I think I said this in the previous PR that clang-format would strip out any indent for this.

File: clang/tools/amdgpu-arch/CMakeLists.txt
Comment (fsfod) [reply clang/tools/amdgpu-arch/CMakeLists.txt:11]: There using target_link_libraries instead of clang_target_link_libraries so there not linked to clang-cpp so we have to disable import macros.

File: clang/include/clang/Support/Compiler.h
Comment (fsfod) [reply clang/include/clang/Support/Compiler.h]: I removed that condition and switched this to being controlled from CMake using CLANG_LINK_CLANG_DYLIB

**Comment by vgvassilev** [comment]
@compnerd, does this PR require more care at that stage or it is good to go?

**Review by compnerd — APPROVED** [review]

**Review by vgvassilev — APPROVED** [review]
Looks good to me, too. We can wait a couple of more days for @AaronBallman to have the final say.

File: clang/include/clang/Support/Compiler.h
Comment (AaronBallman) [inline clang/include/clang/Support/Compiler.h:32]:
```suggestion
// export macros added to them by the clang tool.
```

Comment (AaronBallman) [reply clang/include/clang/Support/Compiler.h:32]: The downside to this is that it's very difficult to know whether or not an API should or shouldn't be exported by the tool. For example, in `Sema`, would we want to export the `ActOn*`  and `Build*` APIs? Probably! But what about the `Check*` APIs? Maybe for most of them. But things like `FillInlineAsmIdentifierInfo`? Probably not.

I seem to recall there being a functional limit to how many interfaces can be exposed in a DLL (I could be remembering wrong though, it's been a while!), so is there harm to over-exposing APIs?

File: clang/include/clang/Support/Compiler.h
Comment (AaronBallman) [inline clang/include/clang/Support/Compiler.h:21]:
```suggestion
/// else that is unannotated having internal visibility.
```

**Review by AaronBallman — APPROVED** [review]
The changes LGTM with some minor changes to the comments. I did have a question, but I don't think it holds up landing the changes.

File: clang/include/clang/Support/Compiler.h
Comment (vgvassilev) [reply clang/include/clang/Support/Compiler.h:32]: To make plugins flexible we probably should export as much as possible. And yes, some of our plugin infrastructure uses even private things from Sema ;)

we could submit patches for them and for some we already have but they cannot be backported across released versions…

Comment (fsfod) [reply clang/include/clang/Support/Compiler.h:32]: I think its only really the LLVM DLL built with MSVC or clang-cl without no export inlines that is sitting close to the symbol export limit.

Comment (AaronBallman) [reply clang/include/clang/Support/Compiler.h:32]: Hmmm, are we close enough to the limit we should be worried now, or are we far enough away from the limit that we can worry in the future if/when we hit it? CC @rnk @compnerd @zmodem for extra opinions on whether we need to think about this more before landing

Comment (zmodem) [reply clang/include/clang/Support/Compiler.h:32]: The limit is 64K exported symbols for a DLL. I don't know how close Clang would get if we'd try to export everything.

Comment (rnk) [reply clang/include/clang/Support/Compiler.h:32]: I don't have specific insight on the number of clang symbols, but for prior art, LLVM has the [`extract_symbols.py`](https://github.com/llvm/llvm-project/blob/main/llvm/utils/extract_symbols.py) script that tries to export everything. You can see there [was an issue in 2023](https://github.com/llvm/llvm-project/issues/60109) where the symbol limit has been reached.

Generally, DLLs with an explicit export list are usually not at risk of hitting the limit, but it's reasonable to ask "how many exports" does the resulting DLL have and compare to 64K.

Comment (fsfod) [reply clang/include/clang/Support/Compiler.h:32]: I managed to get clang-cpp library building without /Zc:dllexportInlines- its at 46748 exported symbols.

Comment (AaronBallman) [reply clang/include/clang/Support/Compiler.h:32]: Would you mind trying an experiment for me? Would you add this to ASTMatchers.h and see how the numbers change for exported symbols:
```
extern const internal::VariadicDynCastAllOfMatcher<Decl, MSPropertyDecl>
    msPropertyDecl;

AST_MATCHER_P(MSPropertyDecl, hasGetter, std::string, Name) {
  return Node.hasGetter() && Node.getGetterId()->isStr(Name);
}
```
I'm asking because AST matchers often have a symbol explosion problem which is easy to overlook. If this adds two exported symbols, I don't worry to much. If it adds 20, it may be more worrying that we're already near 47k out of 65k (not that I think we're going to add that many matchers, just that I worry we'll add more symbols more quickly than we realize).

Comment (fsfod) [reply clang/include/clang/Support/Compiler.h:32]: i did have to export some of the AST matcher variables in another PR https://github.com/llvm/llvm-project/pull/110206.

Comment (fsfod) [reply clang/include/clang/Support/Compiler.h:32]: Just those lines didn't add any more symbols, I assume you meant add definition of the variable to a source file as well. Doing that only added one extra symbol. I should add there were 200 existing symbols with VariadicDynCastAllOfMatcher in there name that were exported.
[astmatcher symbols.txt](https://github.com/user-attachments/files/17346195/astmatcher.symbols.txt)

**Comment by vgvassilev** [comment]
Unless there is more discussion or comments worth addressing I'd propose to merge this PR in the next day or two as it is a bottleneck for a few others in the pipeline.

File: clang/include/clang/Support/Compiler.h
Comment (AaronBallman) [reply clang/include/clang/Support/Compiler.h:32]: Thanks! I am surprised those lines didn't add any more symbols -- they're defined inline in the header files. But since that didn't add a pile of new symbols, I think we're probably okay with the amount of headroom we currently have.

**Review by AaronBallman — APPROVED** [review]
LGTM

**Pull request merged**
